<array>: std::array<T,0> doesn't compile - when type is not default constructible

**Describe the bug**
`std::array<T,0>` doesn't compile - when type has private default constructor, deleted default constructor, or no default constructor. 

If there's a default constructor, it is called once during `std::array<T,0>` construction.

According to LWG-2157, there should be a single element of unspecificed type, and such `array` should be still default constructible.

**Command-line test case**
```
D:\Temp2>type repro.cpp
#include <array>

struct NonDefault {
    NonDefault() = delete;
    NonDefault(int) {}
};

std::array<NonDefault, 0> a;

int main() {
}
D:\Temp2>cl /EHsc /W4 /WX .\repro.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.27.29009.1 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

repro.cpp
.\repro.cpp(8): error C2280: 'std::array<NonDefault,0>::array(void)': attempting to reference a deleted function
c:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\array(770): note: compiler has generated 'std::array<NonDefault,0>::array' here
c:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\array(770): note: 'std::array<NonDefault,0>::array(void)': function was implicitly deleted because a data member invokes a deleted or inaccessible function 'NonDefault::NonDefault(void)'
.\repro.cpp(4): note: 'NonDefault::NonDefault(void)': function was explicitly deleted
```

**Expected behavior**
Test compiles, links and runs.

**STL version**
Microsoft Visual Studio Professional 2019 Preview
Version 16.7.0 Preview 3.1

**Additional context**
 * Skipped libcxx test:
https://github.com/microsoft/STL/blob/06827feb4cdc4d2328dfbfab9fd5302de6058dd9/tests/libcxx/expected_results.txt#L576-L581

This item is also tracked on Developer Community as DevCom-329888, and by Microsoft-internal VSO-207715 / [AB#207715](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/207715).

**vNext note**: Resolving this issue will require breaking binary compatibility. We won't be able to accept pull requests for this issue until the vNext branch is available. See #169 for more information.

--- a/stl/inc/array
+++ b/stl/inc/array
@@ -615,6 +615,8 @@ template <class _First, class... _Rest>
 array(_First, _Rest...) -> array<typename _Enforce_same<_First, _Rest...>::type, 1 + sizeof...(_Rest)>;
 #endif // _HAS_CXX17
 
+struct _Empty_array_element {};
+
 template <class _Ty>
 class array<_Ty, 0> {
 public:
@@ -783,7 +785,7 @@ public:
         _Xout_of_range("invalid array<T, 0> subscript");
     }
 
-    conditional_t<is_default_constructible_v<_Ty>, _Ty, char> _Elems[1];
+    conditional_t<_Is_implicitly_default_constructible<_Ty>::value, _Ty, _Empty_array_element> _Elems[1];
 };
 
 template <class _Ty, size_t _Size, enable_if_t<_Size == 0 || _Is_swappable<_Ty>::value, int> = 0>

--- a/stl/inc/array
+++ b/stl/inc/array
@@ -728,7 +728,7 @@ public:
         _STL_REPORT_ERROR("array subscript out of range");
 #endif // _CONTAINER_DEBUG_LEVEL > 0
 
-        return _Elems[0];
+        return *data();
     }
 
     _NODISCARD const_reference operator[](size_type) const noexcept /* strengthened */ {
@@ -736,7 +736,7 @@ public:
         _STL_REPORT_ERROR("array subscript out of range");
 #endif // _CONTAINER_DEBUG_LEVEL > 0
 
-        return _Elems[0];
+        return *data();
     }
 
     _NODISCARD reference front() noexcept /* strengthened */ {
@@ -744,7 +744,7 @@ public:
         _STL_REPORT_ERROR("array<T, 0>::front() invalid");
 #endif // _CONTAINER_DEBUG_LEVEL > 0
 
-        return _Elems[0];
+        return *data();
     }
 
     _NODISCARD const_reference front() const noexcept /* strengthened */ {
@@ -752,7 +752,7 @@ public:
         _STL_REPORT_ERROR("array<T, 0>::front() invalid");
 #endif // _CONTAINER_DEBUG_LEVEL > 0
 
-        return _Elems[0];
+        return *data();
     }
 
     _NODISCARD reference back() noexcept /* strengthened */ {
@@ -760,7 +760,7 @@ public:
         _STL_REPORT_ERROR("array<T, 0>::back() invalid");
 #endif // _CONTAINER_DEBUG_LEVEL > 0
 
-        return _Elems[0];
+        return *data();
     }
 
     _NODISCARD const_reference back() const noexcept /* strengthened */ {
@@ -768,7 +768,7 @@ public:
         _STL_REPORT_ERROR("array<T, 0>::back() invalid");
 #endif // _CONTAINER_DEBUG_LEVEL > 0
 
-        return _Elems[0];
+        return *data();
     }
 
     _NODISCARD _CONSTEXPR17 _Ty* data() noexcept {
@@ -783,7 +783,7 @@ public:
         _Xout_of_range("invalid array<T, 0> subscript");
     }
 
-    _Ty _Elems[1];
+    conditional_t<is_default_constructible_v<_Ty>, _Ty, char> _Elems[1];
 };
 
 template <class _Ty, size_t _Size, enable_if_t<_Size == 0 || _Is_swappable<_Ty>::value, int> = 0>

--- a/stl/inc/array
+++ b/stl/inc/array
@@ -785,7 +785,9 @@ public:
         _Xout_of_range("invalid array<T, 0> subscript");
     }
 
-    conditional_t<_Is_implicitly_default_constructible<_Ty>::value, _Ty, _Empty_array_element> _Elems[1];
+    conditional_t<disjunction_v<is_default_constructible<_Ty>, _Is_implicitly_default_constructible<_Ty>>, _Ty,
+        _Empty_array_element>
+        _Elems[1];
 };
 
 template <class _Ty, size_t _Size, enable_if_t<_Size == 0 || _Is_swappable<_Ty>::value, int> = 0>

--- a/tests/libcxx/expected_results.txt
+++ b/tests/libcxx/expected_results.txt
@@ -69,6 +69,9 @@ std/language.support/support.limits/support.limits.general/concepts.version.pass
 # Bogus test believes that optional<non_constexpr_destructor> cannot be a literal type
 std/utilities/optional/optional.object/optional.object.dtor/dtor.pass.cpp:0 FAIL
 
+# Bogus test believes that copyability of array<T, 0> must be the same as array<T, 1>
+std/containers/sequences/array/array.cons/implicit_copy.pass.cpp FAIL
+
 
 # *** INTERACTIONS WITH CONTEST / C1XX THAT UPSTREAM LIKELY WON'T FIX ***
 # Tracked by VSO-593630 "<filesystem> Enable libcxx filesystem tests"
@@ -347,13 +350,6 @@ std/input.output/file.streams/fstreams/filebuf.virtuals/underflow.pass.cpp FAIL
 # GH-1295 <array>: array<const T, 0> allows fill() and swap()
 std/containers/sequences/array/array.fill/fill.fail.cpp FAIL
 std/containers/sequences/array/array.swap/swap.fail.cpp FAIL
-
-# GH-942 <array>: std::array<T,0> doesn't compile - when type is not default constructible
-std/containers/sequences/array/array.cons/implicit_copy.pass.cpp FAIL
-std/containers/sequences/array/array.cons/initialization.pass.cpp FAIL
-std/containers/sequences/array/array.data/data_const.pass.cpp FAIL
-std/containers/sequences/array/array.data/data.pass.cpp FAIL
-std/containers/sequences/array/iterators.pass.cpp FAIL
 
 # GH-1006 <algorithm>: debug checks for predicates are observable
 std/algorithms/alg.sorting/alg.merge/inplace_merge_comp.pass.cpp FAIL

--- a/tests/libcxx/skipped_tests.txt
+++ b/tests/libcxx/skipped_tests.txt
@@ -69,6 +69,9 @@ language.support\support.limits\support.limits.general\concepts.version.pass.cpp
 # Bogus test believes that optional<non_constexpr_destructor> cannot be a literal type
 utilities\optional\optional.object\optional.object.dtor\dtor.pass.cpp
 
+# Bogus test believes that copyability of array<T, 0> must be the same as array<T, 1>
+containers\sequences\array\array.cons\implicit_copy.pass.cpp
+
 
 # *** INTERACTIONS WITH CONTEST / C1XX THAT UPSTREAM LIKELY WON'T FIX ***
 # Tracked by VSO-593630 "<filesystem> Enable libcxx filesystem tests"
@@ -347,13 +350,6 @@ input.output\file.streams\fstreams\filebuf.virtuals\underflow.pass.cpp
 # GH-1295 <array>: array<const T, 0> allows fill() and swap()
 containers\sequences\array\array.fill\fill.fail.cpp
 containers\sequences\array\array.swap\swap.fail.cpp
-
-# GH-942 <array>: std::array<T,0> doesn't compile - when type is not default constructible
-containers\sequences\array\array.cons\implicit_copy.pass.cpp
-containers\sequences\array\array.cons\initialization.pass.cpp
-containers\sequences\array\array.data\data_const.pass.cpp
-containers\sequences\array\array.data\data.pass.cpp
-containers\sequences\array\iterators.pass.cpp
 
 # GH-1006 <algorithm>: debug checks for predicates are observable
 algorithms\alg.sorting\alg.merge\inplace_merge_comp.pass.cpp

